homeassistant/components/sensor/history_stats.py

```diff
@@ -0,0 +1,303 @@
+"""
+Component to make instant statistics about your history.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.history_stats/
+"""
+
+import asyncio
+import logging
+import math
+import re
+import time
+
+import voluptuous as vol
+
+import homeassistant.components.history as history
+import homeassistant.components.recorder as recorder
+import homeassistant.helpers.config_validation as cv
+import homeassistant.util.dt as dt_util
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (CONF_NAME, CONF_ENTITY_ID, CONF_STATE)
+from homeassistant.core import callback
+from homeassistant.exceptions import TemplateError
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_state_change
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'history_stats'
+DEPENDENCIES = ['history']
+
+CONF_START = 'start'
+CONF_END = 'end'
+CONF_DURATION = 'duration'
+CONF_PERIOD_KEYS = [CONF_START, CONF_END, CONF_DURATION]
+
+DEFAULT_NAME = 'unnamed statistics'
+UNIT = 'h'
+UNIT_RATIO = '%'
+ICON = 'mdi:chart-line'
+
+PRINT_START = 'from'
+PRINT_END = 'to'
+PRINT_VALUE = 'value'
+PRINT_RATIO = 'ratio'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Required(CONF_STATE): cv.slug,
+    vol.Optional(CONF_START, default=None): cv.template,
+    vol.Optional(CONF_END, default=None): cv.template,
```

Comment:
Because you might want to provide only `start` and `duration`. In that case, `end` should stay undefined.